relayer/cmd/subscribe_beefy.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	log := log.WithField("script", "beefy")
+
+	relaychainEndpoint := config.Relaychain.Endpoint
+	relaychainConn := relaychain.NewConnection(relaychainEndpoint, log)
+	err = relaychainConn.Connect(ctx)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	parachainEndpoint := config.Parachain.Endpoint
+	parachainConn := parachain.NewConnection(parachainEndpoint, nil, log)
+	err = parachainConn.Connect(ctx)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	ch := make(chan interface{})
+
+	log.Info("Subscribing to beefy justifications")
+	sub, err := relaychainConn.GetAPI().Client.Subscribe(context.Background(), "beefy", "subscribeJustifications", "unsubscribeJustifications", "justifications", ch)
+	if err != nil {
+		panic(err)
+	}
+	defer sub.Unsubscribe()
+
+	for {
+		select {
+		case msg := <-ch:
+
+			signedCommitment := &store.SignedCommitment{}
+			err := types.DecodeFromHexString(msg.(string), signedCommitment)
+			if err != nil {
+				log.WithError(err).Error("Failed to decode BEEFY commitment messages")
+			}
+
+			blockNumber := signedCommitment.Commitment.BlockNumber
+
+			log.WithField("commitmentBlockNumber", blockNumber).Info("Witnessed a new BEEFY commitment: \n")
+			if len(signedCommitment.Signatures) == 0 {
+				log.Info("BEEFY commitment has no signatures, skipping...")
+				continue
+			}
+			log.WithField("blockNumber", blockNumber+1).Info("Getting hash for next block")
+			nextBlockHash, err := relaychainConn.GetAPI().RPC.Chain.GetBlockHash(uint64(blockNumber + 1))
+			if err != nil {
+				log.WithError(err).Error("Failed to get block hash")
+			}
+			log.WithField("blockHash", nextBlockHash.Hex()).Info("Got blockhash")
+			GetMMRLeafForBlock(uint64(blockNumber), nextBlockHash, relaychainConn)
+			header, err := GetParaheads(nextBlockHash, relaychainConn)
+			if err != nil {
+				log.WithError(err).Error("Failed to get para heads")
+				return err
+			}
+			GetParaHeadData(header, parachainConn)
+		}
+	}
+}
+
+func GetParaheads(blockHash types.Hash, relaychainConn *relaychain.Connection) (string, error) {
+	type ParaId types.U32
+
+	type ParaHeadsQuery struct {
+		ParaId ParaId
+	}
+
+	query := ParaHeadsQuery{ParaId: 200}
+
+	encoded, err := types.EncodeToBytes(query)
+	if err != nil {
+		log.WithError(err).Error("Error")
+	}
+
+	allParaHeadsStorageKey, err := types.CreateStorageKey(
+		relaychainConn.GetMetadata(),
+		"Paras",
+		"Heads", encoded, nil)
+	if err != nil {
+		log.WithError(err).Error("Failed to create parachain header storage key")
+	}
+
+	// TODO1 - in polkadotjs webapp it seems possible to query this state via paras.heads without specifying any option
+	// and get all parachain heads as a response. eg: trying it on Rococo returns a lot of paraheads.
+	// I think we'll need this (to be able to check against the parachain headers hash
+	// of all headers in MMRLeaf.ParachainHeads), so we should get this working too.
```

Comment:
This is partially resolved in 95964a81adbe57e248040793ef9a2793dd650fb3 . I added the right key by investigating the JSONRPC command sent by polkadot.js apps. This i a bit hacky, since a hardcoded key won't work if metadata ever updates. We should figure out why the computed key is different from the one polkadot.js apps uses.

Reply:
I did notice polkadot.js apps websocket connection makes an extra request for this, something like getKeys before the query for the actual data. Maybe its querying all keys for the mapping first, then appending the key to the base storage key before sending out the RPC?

Reply:
yeh if you try on live Rococo, where there are lots of parachains, it first does state_getKeysPaged then does state_queryStorageAt with a lotttttt of extra params, one for each parachain, basically querying them all together manually. i guess we'll need to do something similar?

Reply:
have a partial fix, will push.

Reply:
working now and gets all headers, though still a small hack to remove some extra bytes from the base key as in https://github.com/Snowfork/polkadot-ethereum/pull/378/files#diff-6bc10d8bbb796a09a8e3772904d4f0950335534c6a6b04d647c24fc66c04b6c8R164 - not sure why

Reply:
Oh, great find! Thanks for diving into this. I'll fix the extra bytes.